controllers/limitador_controller.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2020 Red Hat.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/3scale/limitador-operator/pkg/limitador"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	limitadorv1alpha1 "github.com/3scale/limitador-operator/api/v1alpha1"
+)
+
+// LimitadorReconciler reconciles a Limitador object
+type LimitadorReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=limitador.3scale.net,resources=limitadors,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=limitador.3scale.net,resources=limitadors/status,verbs=get;update;patch
+
+func (r *LimitadorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("limitador", req.NamespacedName)
+
+	// Delete Limitador deployment and service if needed
+	limitadorObj := limitadorv1alpha1.Limitador{}
+	if err := r.Get(context.TODO(), req.NamespacedName, &limitadorObj); err != nil {
+		if errors.IsNotFound(err) {
+			if err = r.ensureLimitadorDeploymentIsDeleted(req.NamespacedName); err != nil {
```

Comment:
An alternative option instead of manually deleting its associated objects you can create those associated objects with an K8s OwnerReference to the Limitador instance. When doing that, when the CR is deleted the associated objects are automatically deleted too.
See https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents for some details on this behavior. If you have some doubts related this possible alternative let me know.

Reply:
I tried to do this. It works when running the operator but fails when running the tests. It seems that the owned references are not deleted even when using the `DeletePropagationForeground` propagation policy. I'd rather solve that in a different PR.

Reply:
ok :+1: